fhirpipe/transform/fhir.py

```diff
@@ -1,122 +1,199 @@
-    if isinstance(structure, dict):
-        # If there are some inputs, we can build the objects to output
-        if "inputs" in structure.keys() and structure["inputs"]:
-            result = build_fhir_leaf_attribute(structure, row, indices)
-            fhir_obj[structure["name"]] = result
 
-        # Otherwise, we can recurse on children
+def fetch_values_from_dataframe(row, mapping_inputs, merging_script):
+    if len(mapping_inputs) == 1:
+        input = mapping_inputs[0]
+        if input["sqlValue"]:
+            sql = input["sqlValue"]
+            column_name = build_col_name(sql["table"], sql["column"], sql["owner"])
+            if input["script"]:
+                column_name = new_col_name(input["script"], column_name)
+            return row[column_name]
         else:
-            is_array = structure["fhirType"] == "array"
-
-            fhir_obj[structure["name"]] = [] if is_array else dict()
+            return input["staticValue"]
 
-            for attr in structure["children"]:
-                if is_array:
-                    create_fhir_list(fhir_obj, attr, structure["name"], row, indices)
-                else:
-                    rec_create_fhir_object(fhir_obj[structure["name"]], attr, row, indices)
+    else:
+        if not merging_script:
+            raise ValueError(
+                "You need to provide a merging script for attributes with several inputs."
+            )
+
+        cols_to_fetch = ([], [])
+        for input in mapping_inputs:
+            # Else retrieve the static value
+            if input["sqlValue"]:
+                sql = input["sqlValue"]
+                column_name = build_col_name(sql["table"], sql["column"], sql["owner"])
+                if input["script"]:
+                    column_name = new_col_name(input["script"], column_name)
+                cols_to_fetch[0].append(column_name)
+
+            # Else retrieve the static value
+            else:
+                cols_to_fetch[1].append(input["staticValue"])
+
+        return row[new_col_name(merging_script, cols_to_fetch)]
+
+
+def handle_array_attributes(attributes_in_array, row):
+    # Check lengths
+    length = 1
+    for attr in attributes_in_array.values():
+        val = fetch_values_from_dataframe(row, attr["inputs"], attr["mergingScript"])
+        if not isinstance(val, tuple) or len(val) == 1:
+            continue
+        assert length == 1 or len(val) == length, "mismatch in array lengths"
+        length = len(val)
+
+    # Now we can build the array
+    array = []
+    for index in range(length):
+        element = build_fhir_object(row, attributes_in_array, index=index)
+        if element:
+            array.append(element)
+
+    return array
+
+
+def insert_in_fhir_object(fhir_object, path, value):
+    if isinstance(value, tuple):
+        assert all(
+            [v == value[0] for v in value]
+        ), "Trying to insert several different values in a non-list attribute"
+        val = value[0]
```

Comment:
Can you explain when this case happens? (having a tuple with all same values)

Reply:
This can happen after a join on a table for which the other values are different and have been squashed. I put a comment